plugins/processors/cumulative_sum/cumulative_sum.go

```diff
@@ -0,0 +1,146 @@
+//go:generate ../../../tools/readme_config_includer/generator
+package cumulative_sum
+
+import (
+	_ "embed"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/config"
+	"github.com/influxdata/telegraf/plugins/processors"
+)
+
+//go:embed sample.conf
+var sampleConfig string
+
+type CumulativeSum struct {
+	Log               telegraf.Logger
+	Fields            []string        `toml:"fields"`
+	DropOriginalField bool            `toml:"drop_original_field"`
+	CleanUpInterval   config.Duration `toml:"clean_up_interval"`
+
+	fieldMap    map[string]bool
+	cache       map[uint64]aggregate
+	nextCleanUp time.Time
+}
+
+type aggregate struct {
+	name       string
+	tags       map[string]string
+	fields     map[string]float64
+	expireTime time.Time
+}
+
+var timeNow = time.Now
+
+func NewCumulativeSum() *CumulativeSum {
+	return &CumulativeSum{
+		DropOriginalField: true,
+		CleanUpInterval:   config.Duration(10 * time.Minute),
+		cache:             make(map[uint64]aggregate),
+	}
+}
+
+func (*CumulativeSum) SampleConfig() string {
+	return sampleConfig
+}
+
+func (c *CumulativeSum) Apply(in ...telegraf.Metric) []telegraf.Metric {
+	c.cleanup()
+	for _, original := range in {
+		id := original.HashID()
+		if _, ok := c.cache[id]; !ok {
+			a := aggregate{
+				name:       original.Name(),
+				tags:       original.Tags(),
+				fields:     make(map[string]float64),
+				expireTime: timeNow().Add(time.Duration(c.CleanUpInterval)),
+			}
+			for _, field := range original.FieldList() {
+				if c.fieldMap != nil {
+					if _, ok := c.fieldMap[field.Key]; !ok {
+						continue
+					}
+				}
+				if fv, ok := convert(field.Value); ok {
+					a.fields[field.Key] = fv
+					original.AddField(field.Key+"_sum", fv)
+					if c.DropOriginalField {
+						original.RemoveField(field.Key)
+					}
+				}
+			}
+			c.cache[id] = a
+		} else {
+			for _, field := range original.FieldList() {
+				if c.fieldMap != nil {
+					if _, ok := c.fieldMap[field.Key]; !ok {
+						continue
+					}
+				}
+				if fv, ok := convert(field.Value); ok {
+					a := c.cache[id]
+					if _, ok := a.fields[field.Key]; !ok {
+						// hit an uncached field of a cached metric
+						a.fields[field.Key] = fv
+					} else {
+						a.fields[field.Key] = a.fields[field.Key] + fv
+					}
+					original.AddField(field.Key+"_sum", a.fields[field.Key])
+					if c.DropOriginalField {
+						original.RemoveField(field.Key)
+					}
+					a.expireTime = timeNow().Add(time.Duration(c.CleanUpInterval))
+					c.cache[id] = a
+				}
+			}
+		}
+	}
+	return in
+}
+
+// Remove expired items from cache
+func (c *CumulativeSum) cleanup() {
+	now := timeNow()
+	if c.nextCleanUp.After(now) {
+		return
+	}
+	c.nextCleanUp = now.Add(time.Duration(c.CleanUpInterval))
+	keep := make(map[uint64]aggregate)
+	for id, a := range c.cache {
+		if a.expireTime.After(now) {
+			keep[id] = a
+		}
+	}
+	c.cache = keep
+}
```

Comment:
What is the use-case for resetting metrics?

Reply:
For mobile metrics, we have a vast tail of outdated tags. For example, now we have app version 1.47 and 99 % of our metrics are with tag `ver=1.47`, but 1 % with tag `ver` other than `1.47` (sometimes it's a `1.26`, for example). We can't ignore that outdated metrics completely, but we would like to avoid saving rare metrics in memory for eternity.

Reply:
But wouldn't it be better to define an `expiry_interval` in the config and if the metric in the cache was not updated for at least that interval it is removed instead of deleting all metrics and destroying the sum of regularly updated metrics?

Reply:
> if the metric in the cache was not updated for at least that interval it is removed instead of deleting all metrics and destroying the sum of regularly updated metrics?

That's precisely what I'm trying to implement here, but with a small optimisation like in [the dedup plugin](https://github.com/influxdata/telegraf/blob/94ddb2b3b499f5d68013bca9199a547dce71591a/plugins/processors/dedup/dedup.go#L28). Running cleaning after each apply will be costly. That's why I added a condition "start cleanup not oftener the cleanup interval".

In other words, the algorithm is:
- check that since the last plugin cleaning us has passed more than the clean up interval
- check each aggregation should it be keeped (last update was less that clean up interval)
- override cache by only keeped values